### PR TITLE
Strict AbstractUser gender

### PR DESCRIPTION
**Pull request by EmanueleMinotto**
Closes https://github.com/elcodi/elcodi/issues/999

### CODE REVIEW

File: src/Elcodi/Component/User/Entity/Abstracts/AbstractUser.php [hunk 1]
```diff
@@ -38,6 +38,21 @@
         DateTimeTrait,
         EnabledTrait;
 
+    const GENDER_UNDEFINED = 0;
```
Comment (mmoreram) [inline src/Elcodi/Component/User/Entity/Abstracts/AbstractUser.php]: Undefined? Can someone have undefined sex? Shouldn't be this null? And being null, this wouldn't be necessary, right?

By default, null (undefined), and each factory should define the default value

File: src/Elcodi/Component/User/Entity/Abstracts/AbstractUser.php [hunk 2]
```diff
@@ -38,6 +38,21 @@
...
+    const GENDER_MALE = 1;
```
Comment (mmoreram) [inline src/Elcodi/Component/User/Entity/Abstracts/AbstractUser.php]: We should add the PHPDoc (like other variables)

File: src/Elcodi/Component/User/Entity/Abstracts/AbstractUser.php [hunk 3]
```diff
@@ -74,11 +89,11 @@
...
 
     /**
-     * @var int
+     * User gender, by default undefined.
      *
-     * gender
+     * @var int
      */
-    protected $gender;
+    protected $gender = self::GENDER_UNDEFINED;
```
Comment (mmoreram) [inline src/Elcodi/Component/User/Entity/Abstracts/AbstractUser.php]: Why do you change the order?

**Comment by mmoreram** [comment]
Thanks @EmanueleMinotto 

We should consider these comments. Furthermore, the default gender should involve as well the factory

File: src/Elcodi/Component/User/Entity/Abstracts/AbstractUser.php (hunk 2 above)
Comment (EmanueleMinotto) [inline src/Elcodi/Component/User/Entity/Abstracts/AbstractUser.php]: :+1:

File: src/Elcodi/Component/User/Entity/Abstracts/AbstractUser.php (hunk 1 above)
Comment (EmanueleMinotto) [inline src/Elcodi/Component/User/Entity/Abstracts/AbstractUser.php]: The undefined value is used for two use cases:
1. https://github.com/elcodi/elcodi/issues/999#issuecomment-160705453 (last point)
2. if a user doesn't want to define their sex, don't know the reason exactly, but I know it's a common thing around the net

File: src/Elcodi/Component/User/Entity/Abstracts/AbstractUser.php (hunk 3 above)
Comment (EmanueleMinotto) [inline src/Elcodi/Component/User/Entity/Abstracts/AbstractUser.php]: the constant is used here to keep the entity consistent, so there's no possibility to have a `null` or a `GENDER_UNDEFINED` value, at least there's the `GENDER_UNDEFINED` value.

**Comment by EmanueleMinotto** [comment]
ping @mmoreram

**Comment by mmoreram** [comment]
@EmanueleMinotto The factory should initialize with anything?

**Comment by EmanueleMinotto** [comment]
I don't see a complex initialization for the `AbstractUser` nor something based on external data, so this attribute could be initializited using default attributes initialization (to keep the state too).
Also to force the initialization from factories, the constructor should be private.

I'm not sure about the factory in this PR :\ no?

**Comment by mmoreram** [comment]
Okay then :)

Could you please rebase the PR?
As soon as is rebased and green, feel free to merge & close :)

Well done dude!

**Comment by EmanueleMinotto** [comment]
@mmoreram rebased :)

**Pull request merged**